src/Git.hs

```diff
@@ -201,100 +199,99 @@ instance FromJSON GitIntegrationFailure
 
 instance ToJSON GitIntegrationFailure where toEncoding = Aeson.genericToEncoding Aeson.defaultOptions
 
-data GitOperationFree a
-  = FetchBranch Branch FetchWithTags a
-  | ForcePush Sha Branch (PushResult -> a)
-  | Push Sha Branch (PushResult -> a)
-  | PushAtomic [SomeRefSpec] (PushResult -> a)
-  | Rebase Sha RemoteBranch (Maybe Sha -> a)
-  | Merge Sha Text (Maybe Sha -> a)
-  | Checkout RemoteBranch (Maybe Sha -> a)
-  | Clone RemoteUrl (CloneResult -> a)
-  | GetParent Sha (Maybe Sha -> a)
-  | RevParse SomeRefSpec (Maybe Sha -> a)
-  | DoesGitDirectoryExist (Bool -> a)
-  | LastTag Sha (Maybe Text -> a)
-  | ShortLog SomeRefSpec SomeRefSpec (Maybe Text -> a)
-  | Tag Sha TagName TagMessage (TagResult -> a)
-  | DeleteTag TagName a
-  | DeleteBranch Branch a
-  | DeleteRemoteBranch Branch (PushResult -> a)
-  | CheckOrphanFixups Sha RemoteBranch (Bool -> a)
-  deriving (Functor)
-
-type GitOperation = Free GitOperationFree
-
-fetchBranch :: Branch -> GitOperation ()
-fetchBranch remoteBranch = liftF $ FetchBranch remoteBranch NoTags ()
-
-fetchBranchWithTags :: Branch -> GitOperation ()
-fetchBranchWithTags remoteBranch = liftF $ FetchBranch remoteBranch WithTags ()
-
-forcePush :: Sha -> Branch -> GitOperation PushResult
-forcePush sha remoteBranch = liftF $ ForcePush sha remoteBranch id
-
-push :: Sha -> Branch -> GitOperation PushResult
-push sha remoteBranch = liftF $ Push sha remoteBranch id
-
-pushAtomic :: [SomeRefSpec] -> GitOperation PushResult
-pushAtomic refs = liftF $ PushAtomic refs id
-
-rebase :: Sha -> RemoteBranch -> GitOperation (Maybe Sha)
-rebase sha ontoBranch = liftF $ Rebase sha ontoBranch id
-
-merge :: Sha -> Text -> GitOperation (Maybe Sha)
-merge sha message = liftF $ Merge sha message id
+data GitOperation :: Effect where
+  FetchBranch :: Branch -> FetchWithTags -> GitOperation m ()
+  ForcePush :: Sha -> Branch -> GitOperation m PushResult
```

Comment:
This shows a major difference, but also a lot of similarities. You still have a data type describing your actions, but it's a GADT instead, and instead of continuations, you just return the thing.

Reply:
Nice, this is way easier to interpret!